### PR TITLE
Docker Extractor

### PR DESCRIPTION
**Pull request by Or-Gabay**
- [x] All [tests](https://ci.appveyor.com/project/jfrog-ecosystem/build-info) passed. If this feature is not already covered by the tests, I added new tests.
-----

### CODE REVIEW

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPull.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    public static void main(String[] ignored) {
+        try {
+            ArtifactoryClientConfiguration clientConfiguration = createArtifactoryClientConfiguration();
+            ArtifactoryDependenciesClientBuilder clientBuilder = new ArtifactoryDependenciesClientBuilder().setClientConfiguration(clientConfiguration, clientConfiguration.publisher);
+            ArtifactoryClientConfiguration.DockerHandler dockerHandler = clientConfiguration.dockerHandler;
+            DockerPull dockerPush = new DockerPull(clientBuilder,
+                    dockerHandler.getImageTag(),
+                    dockerHandler.getHost(),
+                    clientConfiguration.publisher.getUsername(),
+                    clientConfiguration.publisher.getPassword(),
```
Comment (eyalbe4) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPull.java:74]: We should use **clientConfiguration.resolver** and not **clientConfiguration.publisher** in the above two lines.
The same goes for the above line 68.

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    /**
+     * Allow running docker push using a new Java process.
+     */
+    public static void main(String[] ignored) {
+        try {
+            ArtifactoryClientConfiguration clientConfiguration = createArtifactoryClientConfiguration();
+            // Client builders.
+            ArtifactoryBuildInfoClientBuilder buildInfoClientBuilder = new ArtifactoryBuildInfoClientBuilder().setClientConfiguration(clientConfiguration, clientConfiguration.publisher);
+            ArtifactoryDependenciesClientBuilder dependenciesClientBuilder = new ArtifactoryDependenciesClientBuilder().setClientConfiguration(clientConfiguration, clientConfiguration.publisher);
```
Comment (eyalbe4) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java:66]: ArtifactoryDependenciesClientBuilder should not be used for uploads, only for downloads.

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static org.jfrog.build.extractor.packageManager.PackageManagerUtils.createArtifactoryClientConfiguration;
+
+public class DockerPush extends PackageManagerExtractor {
+    ArtifactoryBuildInfoClientBuilder buildInfoClientBuilder;
+    ArtifactoryDependenciesClientBuilder dependenciesClientBuilder;
```
Comment (eyalbe4) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java:24]: ArtifactoryDependenciesClientBuilder --> "dependencies" is for downloads only, not uploads.

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        Properties buildInfoItemsProps = getBuildInfoProps(artifactProperties);
+        String artifactsPropsStr = DockerUtils.buildPropertiesString(artifactProperties);
+        try (ArtifactoryBuildInfoClient buildInfoClient = buildInfoClientBuilder.build();
+             ArtifactoryDependenciesClient dependenciesClient = dependenciesClientBuilder.build()) {
+            Module buildInfoModule = new Module();
+            buildInfoModule.setId(imageTag.substring(imageTag.indexOf("/") + 1));
+            try {
+                findAndSetManifestFromArtifactory(dependenciesClient.getArtifactoryUrl(), dependenciesClient, logger);
+            } catch (IOException e) {
+                // The manifest could be found in Artifactory.
```
Comment (eyalbe4) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:197]: ...could be found... --> ...could not be found...

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerAgentUtils.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+import org.apache.commons.lang.StringUtils;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+
+
+public class DockerAgentUtils {
```
Comment (eyalbe4) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerAgentUtils.java:22]: I think that class can be merged with DockerUtils.
It looks like there's no need to hold both classes.

File: README.md [hunk 6]
```diff
@@ -40,7 +43,7 @@ To build the code without running the tests, add to the "clean build" command th
 ```
 
 ### Testing
-To run *all* tests:   
+To run *all* tests:
```
Comment (eyalbe4) [inline README.md:46]: The README seems to be missing a dedicated test command to run docker tests only, as the other package managers have.

File: build-info-api/src/main/java/org/jfrog/build/api/builder/ArtifactBuilder.java [hunk 7]
```diff
@@ -164,4 +165,12 @@ public ArtifactBuilder addProperty(Object key, Object value) {
         properties.put(key, value);
         return this;
     }
+
+    public ArtifactBuilder addProperty(Map props) {
```
Comment (eyalbe4) [inline build-info-api/src/main/java/org/jfrog/build/api/builder/ArtifactBuilder.java:169]: Why is this method needed?
Should it be named addProperties?
Let's declare the Map it gets with the appropriate key and values types <String, ...>
Let's add a description for it.

The above also goes for the same method in the DependencyBuilder class.

File: build-info-extractor/src/main/java/org/jfrog/build/extractor/clientConfiguration/ArtifactoryClientConfiguration.java [hunk 8]
```diff
@@ -529,6 +528,52 @@ public void setUseDotnetCli(boolean useDotnetCli) {
...
+
+        public String getHost() {
+            return rootConfig.getStringValue(DOCKER_HOST);
+        }
+
+        public void setHost(String host) {
+            rootConfig.setStringValue(DOCKER_HOST, host);
+        }
+
+        public Map<String, String> getArtifactProperties() {
```
Comment (eyalbe4) [inline build-info-extractor/src/main/java/org/jfrog/build/extractor/clientConfiguration/ArtifactoryClientConfiguration.java:552]: The logic of extracting the props from a duplicationstring is duplicated inside the build-info code base. Let's consolidate and reuse it.
The same goes for the below setArtifactProperties method. We should avoid logic duplication.

File: build-info-extractor/src/main/java/org/jfrog/build/extractor/clientConfiguration/ClientConfigurationFields.java [hunk 9]
```diff
@@ -35,6 +35,9 @@
     String PACKAGE_MANAGER_MODULE = "package.manager.module"; // Custom module name for the build-info
     String PIP_ENV_ACTIVATION = "pip.env.activation";
     String DOTNET_USE_DOTNET_CORE_CLI = "dotnet.use.dotnet.core.cli";
+    String DOCKER_IMAGE_TAG = "docker.image.tag";
+    String DOCKER_HOST = "docker.host";
+    String DOCKER_PROPERTIES = "docker.properties";
```
Comment (eyalbe4) [inline build-info-extractor/src/main/java/org/jfrog/build/extractor/clientConfiguration/ClientConfigurationFields.java:40]: What is String "docker.properties"? Isn't this just properties we attach to artifacts? If so, why can't we use the usual (generic) way fo attaching properties to artifacts?

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java (hunk 2 above)
Comment (Or-Gabay) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java:66]: We use it to download the images' manifest

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java (hunk 3 above)
Comment (Or-Gabay) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java:24]: https://github.com/jfrog/build-info/pull/373#discussion_r471093629

File: build-info-extractor/src/main/java/org/jfrog/build/extractor/clientConfiguration/ClientConfigurationFields.java (hunk 9 above)
Comment (Or-Gabay) [reply build-info-extractor/src/main/java/org/jfrog/build/extractor/clientConfiguration/ClientConfigurationFields.java:40]: Removed.

File: README.md [hunk 10]
```diff
@@ -89,6 +88,20 @@ When running on a unix machine, provide the path to the 'bin' directory.
 > ./gradlew clean build-info-extractor-nuget:test
 ```
 
+#### Docker tests
+* Docker tests run only on Linux/mac.
+* In addition to general environment variables you must set the following docker environment variables:
```
Comment (eyalbe4) [inline README.md:93]: In addition to general environment variables you must set the following docker environment variables:
-->
In addition to the general environment variables required for running the tests, you must set the following environment variables, required for the docker tests:

File: build-info-api/src/main/java/org/jfrog/build/api/util/Utils.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+public class Utils {
+    /**
+     * Build string of properties in the following form:
+     * key=value;key=value...
+     *
+     * @param properties - keys and values to concat
+     * @return
+     */
+    public static String buildPropertiesString(ArrayListMultimap<String, String> properties) {
```
Comment (eyalbe4) [inline build-info-api/src/main/java/org/jfrog/build/api/util/Utils.java:17]: I could find at least one more place in the code where this logic is duplicated. See the buildMatrixParamsString method in the DeploymentUrlUtils class.

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerAgentUtils.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+import com.github.dockerjava.core.command.PushImageResultCallback;
+import com.github.dockerjava.netty.NettyDockerCmdExecFactory;
+import org.apache.commons.lang.StringUtils;
+import org.jfrog.build.api.util.Log;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
```
Comment (eyalbe4) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerAgentUtils.java:20]: This import seems to be redundant now.

File: build-info-api/src/main/java/org/jfrog/build/api/util/Utils.java (hunk 11 above)
Comment (Or-Gabay) [reply build-info-api/src/main/java/org/jfrog/build/api/util/Utils.java:17]: `buildMatrixParams` returns a different result that corrupts the `vcs.url`.

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerAgentUtils.java (hunk 12 above)
Comment (Or-Gabay) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerAgentUtils.java:20]: it is done automatically by IntelliJ formatting.

Comment (eyalbe4) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerAgentUtils.java:20]: You should have removed this import after stopping to use the Map.

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerAgentUtils.java (hunk 5 above)
Comment (Or-Gabay) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerAgentUtils.java:22]: Renamed to `DockerJavaWrapper`

File: build-info-api/src/main/java/org/jfrog/build/api/util/Utils.java (hunk 11 above)
Comment (Or-Gabay) [reply build-info-api/src/main/java/org/jfrog/build/api/util/Utils.java:17]: buildMatrixParams removed.

**Pull request merged**